nanshe_ipython.ipynb

```diff
@@ -328,6 +331,261 @@
+        "\n",
+        "        # Determine phase shifts\n",
+        "        J = da_imgs_fft.dtype.type(1j)\n",
+        "        indices = cartesian_product(da_imgs_flt.shape[1:])\n",
+        "        fourier_unit_shift = indices * -wave_vector\n",
+        "        fourier_unit_shift = fourier_unit_shift.T.reshape(\n",
+        "            (da_imgs_flt.ndim - 1,) + da_imgs_flt.shape[1:]\n",
+        "        )\n",
+        "        fourier_unit_shift = fourier_unit_shift.persist()\n",
+        "\n",
+        "        # Initialize\n",
+        "        i = 0\n",
+        "        avg_rel_dist = 1.0\n",
+        "        tmpl_hist_wght = da_imgs_flt.dtype.type(tmpl_hist_wght)\n",
+        "        shifts = da.zeros(\n",
+        "            (len(da_imgs_flt), da_imgs_flt.ndim - 1),\n",
+        "            dtype=int,\n",
+        "            chunks=(1, da_imgs_flt.ndim - 1)\n",
+        "        )\n",
+        "        da_imgs_fft_tmplt = da_imgs_fft.mean(axis=0).persist()\n",
+        "\n",
+        "        while avg_rel_dist > thld_rel_dist and i < num_reps:\n",
+        "            # Compute the template FFT\n",
+        "            phase_shifts = da.exp(\n",
+        "                J * da.tensordot(shifts, fourier_unit_shift, axes=1)\n",
+        "            )\n",
+        "            shifted_frames = da_imgs_fft * phase_shifts\n",
+        "            da_imgs_fft_tmplt = (\n",
+        "                tmpl_hist_wght * da_imgs_fft_tmplt +\n",
+        "                (1 - da_imgs_fft_tmplt) * shifted_frames.mean(axis=0)\n",
+        "            )\n",
+        "            da_imgs_fft_tmplt = da_imgs_fft_tmplt.rechunk(da_imgs_fft_tmplt.shape)\n",
+        "            da_imgs_fft_tmplt = da_imgs_fft_tmplt.persist()\n",
+        "\n",
+        "            # Find the best overlap with the template.\n",
+        "            overlap = da.absolute(ifftn(\n",
+        "                da_imgs_fft * da_imgs_fft_tmplt[None].conj(),\n",
+        "                axes=tuple(irange(1, imgs.ndim))\n",
+        "            ))\n",
+        "            overlap_max = overlap.max(axis=tuple(irange(1, imgs.ndim)))\n",
+        "            overlap_max_match = (overlap == overlap_max[(Ellipsis,) + (None,) * (imgs.ndim - 1)])\n",
+        "\n",
+        "            # Compute the shift for each frame.\n",
+        "            old_shifts = shifts\n",
+        "            shifts = []\n",
+        "            for j in irange(len(overlap_max_match)):\n",
+        "                shift_j = dask.delayed(compute_offset)(overlap_max_match[j])\n",
+        "                shift_j = da.from_delayed(shift_j, (2,), int)\n",
+        "                shifts.append(shift_j)\n",
+        "            shifts = da.stack(shifts)\n",
+        "\n",
+        "            # Remove any collective frame drift.\n",
+        "            drift = shifts.mean(axis=0).round().astype(shifts.dtype)\n",
+        "            shifts = shifts - drift\n",
```

Comment:
Not sure it is necessary, but maybe it should be `drift[None]`. Alternatively the `mean` call above could be provided `keepdims=True`.